### ISSUE TITLE
XGBClassifier fit() custom eval_metric is not called

### ISSUE THREAD

**Issue by jonyer** [dmlc/xgboost#5786]
I am passing a custom evaluation metric function to XGBClassifier fit(), but the Python API discards callable (i.e., custom) functions. Why? It also seems to discard any string inputs. Regardless of what I pass in, the results are the same. I tried all objective functions as well, and the eval_metric parameter does not have an impact on any of them. 

How are we supposed to pass in custom functions? Or is this functionality not supported by the Python Scikit API?

model = XGBClassifier(scale_pos_weight=scale_pos_weight)
model.fit(X_train, y_train, eval_metric=xgb_f1)

**Comment by hcho3** [comment]
@jonyer 

> I am passing a custom evaluation metric function to XGBClassifier fit(), but the Python API discards callable (i.e., custom) functions.

It works for me. Here is my program:
```python
from typing import Tuple, Dict, List

import numpy as np
import xgboost as xgb
from sklearn.model_selection import train_test_split

# shape of generated data.
kRows = 4096
kCols = 16

kRatio = 0.7
kSeed = 1994

kBoostRound = 20

np.random.seed(seed=kSeed)

def generate_data() -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    X = np.random.randn(kRows, kCols)
    y = np.random.randn(kRows)
    y += np.abs(np.min(y))

    # rmsle requires all label be greater than -1.
    assert np.all(y > -1.0)

    train_portion = int(kRows * kRatio)
    
    X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=kRatio, random_state=kSeed)
    return X_train, X_test, y_train, y_test

def rmsle(predt: np.ndarray, dtrain: xgb.DMatrix) -> Tuple[str, float]:
    y = dtrain.get_label()
    predt[predt < -1] = -1 + 1e-6
    elements = np.power(np.log1p(y) - np.log1p(predt), 2)
    return 'PyRMSLE', float(np.sqrt(np.sum(elements) / len(y)))


def main():
    X_train, X_test, y_train, y_test = generate_data()
    clf = xgb.XGBRegressor(n_estimators=kBoostRound, objective='reg:squarederror',
            tree_method='hist', seed=kSeed)
    clf.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_test, y_test)],
            eval_metric=rmsle)

if __name__ == '__main__':
    main()
```

Result (XGBoost 1.1.1):
```
[0]	validation_0-rmse:2.33321	validation_1-rmse:2.35009	validation_0-PyRMSLE:0.64873	validation_1-PyRMSLE:0.65055
[1]	validation_0-rmse:1.75676	validation_1-rmse:1.80821	validation_0-PyRMSLE:0.43219	validation_1-PyRMSLE:0.44691
[2]	validation_0-rmse:1.38112	validation_1-rmse:1.46679	validation_0-PyRMSLE:0.31647	validation_1-PyRMSLE:0.34162
[3]	validation_0-rmse:1.12578	validation_1-rmse:1.27071	validation_0-PyRMSLE:0.24837	validation_1-PyRMSLE:0.29057
[4]	validation_0-rmse:0.97474	validation_1-rmse:1.16187	validation_0-PyRMSLE:0.21377	validation_1-PyRMSLE:0.26594
[5]	validation_0-rmse:0.87860	validation_1-rmse:1.10463	validation_0-PyRMSLE:0.19428	validation_1-PyRMSLE:0.25501
[6]	validation_0-rmse:0.81496	validation_1-rmse:1.07550	validation_0-PyRMSLE:0.18310	validation_1-PyRMSLE:0.25030
[7]	validation_0-rmse:0.76317	validation_1-rmse:1.06308	validation_0-PyRMSLE:0.17378	validation_1-PyRMSLE:0.24899
[8]	validation_0-rmse:0.73728	validation_1-rmse:1.05851	validation_0-PyRMSLE:0.17022	validation_1-PyRMSLE:0.24892
[9]	validation_0-rmse:0.69912	validation_1-rmse:1.06052	validation_0-PyRMSLE:0.16307	validation_1-PyRMSLE:0.25004
[10]	validation_0-rmse:0.66966	validation_1-rmse:1.06130	validation_0-PyRMSLE:0.15724	validation_1-PyRMSLE:0.25068
[11]	validation_0-rmse:0.63564	validation_1-rmse:1.06292	validation_0-PyRMSLE:0.15040	validation_1-PyRMSLE:0.25134
[12]	validation_0-rmse:0.60166	validation_1-rmse:1.06753	validation_0-PyRMSLE:0.14349	validation_1-PyRMSLE:0.25270
[13]	validation_0-rmse:0.58108	validation_1-rmse:1.06794	validation_0-PyRMSLE:0.13882	validation_1-PyRMSLE:0.25289
[14]	validation_0-rmse:0.56392	validation_1-rmse:1.07069	validation_0-PyRMSLE:0.13501	validation_1-PyRMSLE:0.25365
[15]	validation_0-rmse:0.54285	validation_1-rmse:1.07463	validation_0-PyRMSLE:0.12979	validation_1-PyRMSLE:0.25452
[16]	validation_0-rmse:0.52637	validation_1-rmse:1.07646	validation_0-PyRMSLE:0.12632	validation_1-PyRMSLE:0.25493
[17]	validation_0-rmse:0.50439	validation_1-rmse:1.07910	validation_0-PyRMSLE:0.12157	validation_1-PyRMSLE:0.25542
[18]	validation_0-rmse:0.48786	validation_1-rmse:1.08087	validation_0-PyRMSLE:0.11788	validation_1-PyRMSLE:0.25576
[19]	validation_0-rmse:0.47791	validation_1-rmse:1.08157	validation_0-PyRMSLE:0.11571	validation_1-PyRMSLE:0.25587
```

**Comment by jonyer** [comment]
This works for me too. I wasn't aware that you had to specify eval_set when specifying eval_metric. I have not seen that in the documentation. When I add eval_set, my custom function is called.

However, regardless of what my function returns (i.e., "F1 score" or "1 - F1 score" or even just "0"), the optimization result is the same. 
Here are the results of "F1 score" and "1 - F1 score" and "0". As you can see, the classification report is exactly the same for all of them.
Also, I'm using XGBClassifier, no Regressor, in case that makes a difference.

[99]	validation_0-error:0.32139	validation_1-error:0.36103	validation_0-f1:0.67313	validation_1-f1:0.74420

              precision    recall  f1-score   support

           0       0.92      0.69      0.79      9555
           1       0.23      0.59      0.33      1466

    accuracy                           0.68     11021
   macro avg       0.57      0.64      0.56     11021
weighted avg       0.82      0.68      0.73     11021

              precision    recall  f1-score   support

           0       0.91      0.66      0.76      2421
           1       0.17      0.51      0.26       335

    accuracy                           0.64      2756
   macro avg       0.54      0.58      0.51      2756
weighted avg       0.82      0.64      0.70      2756


---------------------------------------------------
[99]	validation_0-error:0.32139	validation_1-error:0.36103	validation_0-f1:0.32687	validation_1-f1:0.25580

              precision    recall  f1-score   support

           0       0.92      0.69      0.79      9555
           1       0.23      0.59      0.33      1466

    accuracy                           0.68     11021
   macro avg       0.57      0.64      0.56     11021
weighted avg       0.82      0.68      0.73     11021


              precision    recall  f1-score   support

           0       0.91      0.66      0.76      2421
           1       0.17      0.51      0.26       335

    accuracy                           0.64      2756
   macro avg       0.54      0.58      0.51      2756
weighted avg       0.82      0.64      0.70      2756


---------------------------------------------------
[99]	validation_0-error:0.32139	validation_1-error:0.36103	validation_0-f1:0.00000	validation_1-f1:0.00000

              precision    recall  f1-score   support

           0       0.92      0.69      0.79      9555
           1       0.23      0.59      0.33      1466

    accuracy                           0.68     11021
   macro avg       0.57      0.64      0.56     11021
weighted avg       0.82      0.68      0.73     11021


              precision    recall  f1-score   support

           0       0.91      0.66      0.76      2421
           1       0.17      0.51      0.26       335

    accuracy                           0.64      2756
   macro avg       0.54      0.58      0.51      2756
weighted avg       0.82      0.64      0.70      2756

**Comment by hcho3** [comment]
> regardless of what my function returns (i.e., "F1 score" or "1 - F1 score" or even just "0"), the optimization results are the same.

That’s expected. We are defining a custom metric, and a custom metric has no effect on the optimization. To modify optimization, should define a **custom objective** instead.

**Comment by hcho3** [comment]
Here is an example of using both custom objective and custom metric:
```python
from typing import Tuple, Dict, List

import numpy as np
import xgboost as xgb
from sklearn.model_selection import train_test_split

# shape of generated data.
kRows = 4096
kCols = 16

kOutlier = 10000                # mean of generated outliers
kNumberOfOutliers = 64

kRatio = 0.7
kSeed = 1994

kBoostRound = 20

np.random.seed(seed=kSeed)

def generate_data() -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    X = np.random.randn(kRows, kCols)
    y = np.random.randn(kRows)
    y += np.abs(np.min(y))

    # Create outliers
    for i in range(0, kNumberOfOutliers):
        ind = np.random.randint(0, len(y) - 1)
        y[ind] += np.random.randint(0, kOutlier)

    # rmsle requires all label be greater than -1.
    assert np.all(y > -1.0)

    train_portion = int(kRows * kRatio)
    
    X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=kRatio, random_state=kSeed)
    return X_train, X_test, y_train, y_test

def squared_log_obj(y_true: np.ndarray,
                    y_pred: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
    y_pred[y_pred < -1] = -1 + 1e-6
    grad = (np.log1p(y_pred) - np.log1p(y_true)) / (y_pred + 1)
    hess = ((-np.log1p(y_pred) + np.log1p(y_true) + 1) / np.power(y_pred + 1, 2))
    return grad, hess

def rmsle(y_pred: np.ndarray, dtrain: xgb.DMatrix) -> Tuple[str, float]:
    y_true = dtrain.get_label()
    y_pred[y_pred < -1] = -1 + 1e-6
    elements = np.power(np.log1p(y_true) - np.log1p(y_pred), 2)
    return 'PyRMSLE', float(np.sqrt(np.sum(elements) / len(y_true)))

def main():
    X_train, X_test, y_train, y_test = generate_data()
    # Train with reg:squarederror objective
    clf = xgb.XGBRegressor(n_estimators=kBoostRound, objective='reg:squarederror',
            tree_method='hist', seed=kSeed)
    clf.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_test, y_test)],
            eval_metric=rmsle)
    # Train with custom objective
    clf = xgb.XGBRegressor(n_estimators=kBoostRound, objective=squared_log_obj,
            tree_method='hist', seed=kSeed)
    clf.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_test, y_test)],
            eval_metric=rmsle)

if __name__ == '__main__':
    main()
```

**Comment by shuhaojie** [comment]
> def squared_log_obj(y_true: np.ndarray,
>                     y_pred: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
>     y_pred[y_pred < -1] = -1 + 1e-6
>     grad = (np.log1p(y_pred) - np.log1p(y_true)) / (y_pred + 1)
>     hess = ((-np.log1p(y_pred) + np.log1p(y_true) + 1) / np.power(y_pred + 1, 2))
>     return grad, hess
> 
What is the meaning of `grad` and `hess` here? If I'd like to custom a loss function, is it necessary to return `grad` and `hess`?
